annotation/tests/test_tasks_services.py

```diff
@@ -0,0 +1,343 @@
+):
+    with pytest.raises(HTTPException) as excinfo:
+        validate_user_actions(
+            is_validation=True,
+            failed=failed,
+            annotated=annotated,
+            not_processed=set(),
+            annotation_user=annotation_user,
+            validation_user=validation_user,
+        )
+    assert re.match(expected_error_message_pattern, excinfo.value.detail)
+
+
+@pytest.mark.parametrize(
+    (
+        "failed",
+        "annotated",
+        "not_processed",
+        "expected_error_message_pattern",
+    ),
+    (
+        (
+            set(),
+            set(),
+            set(),
+            r"Validator did not mark any pages as failed",
+        ),
+        (
+            {1},
+            set(),
+            set(),
+            r"Validator did not edit any pages",
+        ),
+        (
+            {1},
+            {2},
+            {3},
+            r"Cannot finish validation task. There are not processed pages",
+        ),
+    ),
+)
+def test_validate_user_actions_invalid_states(
+    failed: set,
+    annotated: set,
+    not_processed: set,
+    expected_error_message_pattern: str,
+):
+    with pytest.raises(HTTPException) as excinfo:
+        validate_user_actions(
+            is_validation=True,
+            failed=failed,
+            annotated=annotated,
+            not_processed=not_processed,
+            annotation_user=True,
+            validation_user=True,
+        )
+    assert re.match(expected_error_message_pattern, excinfo.value.detail)
+
+
+def test_create_annotation_task_success():
+
+    with patch("sqlalchemy.orm.Session", spec=True) as mock_session, patch(
+        "annotation.jobs.services.update_user_overall_load"
+    ) as mock_update_user_overall_load, patch(
+        "annotation.models.ManualAnnotationTask"
+    ), patch(
+        "annotation.schemas.tasks.ManualAnnotationTaskInSchema"
+    ):
+
+        db_session = mock_session()
+        create_annotation_task(
+            db_session,
+            ManualAnnotationTaskInSchema(
+                file_id=1,
+                pages={1, 2},
+                job_id=2,
+                user_id=uuid4(),
+                is_validation=True,
+                deadline=None,
+            ),
+        )
+
+        assert db_session.add.call_count == 2
+        mock_update_user_overall_load.assert_not_called()
+        db_session.commit.assert_called_once()
+
+
+def test_create_annotation_task_commit_failure():
+
+    with patch("sqlalchemy.orm.Session", spec=True) as mock_session, patch(
+        "annotation.jobs.services.update_user_overall_load"
+    ) as mock_update_user_overall_load:
+        db_session = mock_session()
+        db_session.commit.side_effect = Exception("Commit failed")
+
+        with pytest.raises(Exception):
+            create_annotation_task(db_session, None)
+
+        db_session.commit.assert_not_called()
+        mock_update_user_overall_load.assert_not_called()
```

Comment:
In this test we actually aren't testing an expected exception of this function. We are mocking the side_effect with an exception that the function doesn't declare in its code.